src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,289 @@
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var model = await _documentSymbolQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            var searchText = viewModelStateData.SelectLastNonNullOrDefault(static x => x.SearchText);
+            var position = viewModelStateData.SelectLastNonNullOrDefault(static x => x.CaretPositionOfNodeToSelect);
+            var expansion = viewModelStateData.SelectLastNonNullOrDefault(static x => x.ExpansionOption);
+            var dataUpdated = viewModelStateData.Any(static x => x.DataUpdated);
+
+            // These updates always require a valid model to perform
+            if (model is not null)
+            {
+                if (string.IsNullOrEmpty(searchText) && dataUpdated)
+                {
+                    var documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                    ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (searchText is { } currentQuery)
+                {
+                    ImmutableArray<DocumentSymbolDataViewModel> documentSymbolViewModelItems;
+                    if (currentQuery == string.Empty)
+                    {
+                        // search was cleared, show all data.
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(model.DocumentSymbolData);
+                        ApplyExpansionStateToNewItems(documentSymbolViewModelItems, DocumentSymbolViewModelItems);
+                    }
+                    else
+                    {
+                        // We are going to show results so we unset any expand / collapse state
+                        documentSymbolViewModelItems = DocumentOutlineHelper.GetDocumentSymbolItemViewModels(
+                             DocumentOutlineHelper.SearchDocumentSymbolData(model.DocumentSymbolData, currentQuery, cancellationToken));
+                    }
+
+                    DocumentSymbolViewModelItems = documentSymbolViewModelItems;
+                }
+
+                if (position is { } currentPosition)
+                {
+                    var caretPoint = currentPosition.Point.GetPoint(_textBuffer, PositionAffinity.Predecessor);
+                    if (!caretPoint.HasValue)
+                    {
+                        // document has changed since we were queued to the point that we can't find this position.
+                        return;
+                    }
+
+                    var symbolToSelect = DocumentOutlineHelper.GetDocumentNodeToSelect(DocumentSymbolViewModelItems, model.OriginalSnapshot, caretPoint.Value);
+                    if (symbolToSelect is null)
+                    {
+                        return;
```

Comment:
so all these returns are a bit suspect to me, as i can't tell if it's intentional that you're not calling into the if-statemetns on line 261 and beyond.